File: app/src/main/java/com/skybox/seven/covid/data/AppDatabase.java
package com.skybox.seven.covid.data;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.skybox.seven.covid.BuildConfig;
import com.skybox.seven.covid.data.daos.AdviceDAO;
import com.skybox.seven.covid.data.daos.InfoGraphicDAO;
import com.skybox.seven.covid.data.daos.LanguageDAO;
import com.skybox.seven.covid.data.daos.MythsDAO;
import com.skybox.seven.covid.data.daos.QnADAO;
import com.skybox.seven.covid.data.daos.SelfTestAnswerDAO;
import com.skybox.seven.covid.data.daos.SelfTestCompleteDAO;
import com.skybox.seven.covid.data.daos.SelfTestQuestionDAO;
import com.skybox.seven.covid.data.entities.Advice;
import com.skybox.seven.covid.data.entities.InfoGraphic;
import com.skybox.seven.covid.data.entities.Language;
import com.skybox.seven.covid.data.entities.Myth;
import com.skybox.seven.covid.data.entities.Qna;
import com.skybox.seven.covid.data.entities.SelfTestAnswer;
import com.skybox.seven.covid.data.entities.SelfTestComplete;
import com.skybox.seven.covid.data.entities.SelfTestQuestion;
import com.skybox.seven.covid.helpers.RoomTypeConverters;
import com.skybox.seven.covid.util.DataBaseInitWorker;

import java.util.concurrent.Executors;

@Database(entities = {SelfTestAnswer.class, SelfTestQuestion.class, SelfTestComplete.class, Myth.class, Advice.class, InfoGraphic.class, Language.class, Qna.class}, version = BuildConfig.HEALTH_DB_VERSION)
@TypeConverters({RoomTypeConverters.class})
public abstract class AppDatabase extends RoomDatabase {
    private static final String TAG = "AppDatabase";
    private static AppDatabase INSTANCE;
    private final static String DB_NAME = "MalawiHealth";

    public abstract SelfTestAnswerDAO selfTestAnswerDAO();
    public abstract SelfTestQuestionDAO selfTestQuestionDAO();
    public abstract SelfTestCompleteDAO selfTestCompleteDAO();
    public abstract MythsDAO mythsDAO();
    public abstract LanguageDAO languageDAO();
    public abstract AdviceDAO adviceDAO();
    public abstract InfoGraphicDAO infoGraphicDAO();
    public abstract QnADAO qnADAO();

    public static AppDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                INSTANCE = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, DB_NAME)
                        .allowMainThreadQueries()
                        .addCallback(new Callback() {
                            @Override
                            public void onCreate(@NonNull SupportSQLiteDatabase db) {
                                super.onCreate(db);
                                Executors.newSingleThreadExecutor().execute(() -> {
                                    OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(DataBaseInitWorker.class).build();
                                    WorkManager.getInstance(context).enqueue(workRequest);
                                });
                            }
                        })
                        .build();
            }
        }
        return INSTANCE;
    }

    public static AppDatabase initStuff (final Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                INSTANCE = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, DB_NAME)
                        .allowMainThreadQueries()
                        .addCallback(new Callback() {
                            @Override
                            public void onCreate(@NonNull SupportSQLiteDatabase db) {
                                super.onCreate(db);
                                Log.e(TAG, "onCreating");
                                Executors.newSingleThreadExecutor().execute(() -> {
                                    OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(DataBaseInitWorker.class).build();
                                    WorkManager.getInstance(context).enqueue(workRequest);
                                });
                            }
                        })
                        .build();
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/skybox/seven/covid/data/AppDatabase.kt
package com.skybox.seven.covid.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.work.*
import com.skybox.seven.covid.BuildConfig
import com.skybox.seven.covid.data.daos.*
import com.skybox.seven.covid.data.entities.*
import com.skybox.seven.covid.helpers.RoomTypeConverters
import com.skybox.seven.covid.work.DataBaseInitWorker
import com.skybox.seven.covid.work.GrabSelfTestsWorker
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

@Database(entities = [SelfTestAnswer::class, SelfTestQuestion::class, SelfTestComplete::class, Myth::class, Advice::class, InfoGraphic::class, Language::class, Qna::class], version = BuildConfig.HEALTH_DB_VERSION)
@TypeConverters(RoomTypeConverters::class)
abstract class AppDatabase : RoomDatabase() {
    abstract fun selfTestAnswerDAO(): SelfTestAnswerDAO
    abstract fun selfTestQuestionDAO(): SelfTestQuestionDAO
    abstract fun selfTestCompleteDAO(): SelfTestCompleteDAO
    abstract fun mythsDAO(): MythsDAO
    abstract fun languageDAO(): LanguageDAO
    abstract fun adviceDAO(): AdviceDAO
    abstract fun infoGraphicDAO(): InfoGraphicDAO
    abstract fun qnADAO(): QnADAO

    companion object {
        private const val DB_NAME = "MalawiHealth"
        @JvmStatic
        fun getDatabase(context: Context): AppDatabase =Room.databaseBuilder(context.applicationContext, AppDatabase::class.java, DB_NAME)
                            .allowMainThreadQueries()
                            .addCallback(object : Callback() {
                                override fun onCreate(db: SupportSQLiteDatabase) {
                                    super.onCreate(db)
                                    Executors.newSingleThreadExecutor().execute {
                                        val workRequest = OneTimeWorkRequest.Builder(DataBaseInitWorker::class.java).build()

                                        val constraints = Constraints.Builder()
                                                .setRequiredNetworkType(NetworkType.CONNECTED)
                                                .build()

                                        val selfTest = OneTimeWorkRequest.Builder(GrabSelfTestsWorker::class.java)
                                                .addTag(GrabSelfTestsWorker.TAG)
                                                .setConstraints(constraints)
                                                .setBackoffCriteria(
                                                        BackoffPolicy.LINEAR,
                                                        OneTimeWorkRequest.MIN_BACKOFF_MILLIS,
                                                        TimeUnit.MILLISECONDS
                                                )
                                                .build()
                                        WorkManager.getInstance(context).enqueue(arrayListOf(workRequest, selfTest))
                                    }
                                }
                            })
                            .build()
                }
}